### ISSUE TITLE
Azure Storage Explorer is leaking X11 clients

### ISSUE THREAD

**Issue by GeeWee** [microsoft/AzureStorageExplorer#3376]
**Storage Explorer Version**: 1.14.0
**Build Number**: 20200624.3
**Platform/OS**: Fedora 32
**Architecture**: x64
**Regression From**: Potentially a regression. It seems to be a new bug that I only started encountering over the last few months.

## Bug Description ##
Azure Storage Explorer leaks Xorg connections. This means that after using it for a little while, the xorg connections will continue to rise, until the maximum connections have been reached, and everything stops working.

Closing down Azure Storage Explorer restores the system to a usable state.

I'm not particularly well versed in Unix and Gnome.
I used the following issues to debug:
https://github.com/electron/electron/issues/2873
https://github.com/electron/electron/issues/2922

### To demonstrate the amount of sockets open.
Running `lsof -U | tr -s ' ' | cut -d ' ' -f 1 | sort | uniq -c | sort -rns` - which will list the amounts of processes with open files/sockets.

Right after starting Storage Explorer
```
140 dbus-brok
135 StorageEx
88 code
84 firefox
83 chrome
...
```
After clicking around a bit
```
158 StorageEx
134 dbus-brok
88 code
```
and even later
```
189 StorageEx
132 dbus-brok
```
etc.

### Journalctl log
When the system is in this state, Azure Storage explorer continues to try and create clients.
This is from the journalctl log:

```
Aug 11 09:49:27 localhost.localdomain storage-explorer_storage-explorer.desktop[26319]: Maximum number of clients reachedFailed to setup connection to X server.
Aug 11 09:49:27 localhost.localdomain storage-explorer_storage-explorer.desktop[26319]: Maximum number of clients reachedFailed to setup connection to X server.
Aug 11 09:49:28 localhost.localdomain storage-explorer_storage-explorer.desktop[26319]: Maximum number of clients reachedFailed to setup connection to X server.
Aug 11 09:49:28 localhost.localdomain storage-explorer_storage-explorer.desktop[26319]: Maximum number of clients reachedFailed to setup connection to X server.
... (and many more)
```

## Steps to Reproduce ##
- Start Azure Storage Explorer
- Click around 20-30 times
- See that the open sockets keep increasing
- Click around 20-30 times more
- System stops working. New programs can't be started etc.
- Close Azure Storage Explorer
- System can now open new programs.

**Comment by JasonYeMSFT** [comment]
@GeeWee Could you try upgrading to 1.14.2 to see if it fix the problem. I have checked in a fix to properly close XServer connections in our 1.14.1 hotfix.

Related to #3206.

**Comment by GeeWee** [comment]
It does fix the overarching issue. I didn't try upgrading because I didn't see anything in the patch notes.

The socket usage is still very high, and it does rise, (I've managed to get it up to 281)
```
281 StorageEx
120 dbus-brok
120 MainThrea
108 Web\x20Co
81 teams
...
```
It seems to open new sockets when opening new functionality, and when closing all tabs it goes down to 229 again. So the socket usage is still high, but it doesn't grow uncontrollably anymore.
Depending on whether or not you want to investigate this further, you can close this issue. Thanks!

**Comment by MRayermannMSFT** [comment]
@GeeWee 

> I didn't try upgrading because I didn't see anything in the patch notes

Sorry about that, we had a hard time 100% verifying the fix, and the reporter of the other issue never confirmed if the fix worked for him. We tend to only put things in the release notes if we know they fix the issue.

I think that viewpoint might be worth reconsidering. We shouldn't be so afraid to say "There's this issue and we think we've fixed it, please let us know if you still see it". That would've helped you faster than how we originally handled things.

> The socket usage is still very high

At what point do you consider the usage high? I'd love to have a target to aim for. 

> It seems to open new sockets when opening new functionality, and when closing all tabs it goes down to 229 again.

For what resources types are you opening tabs? Blobs, disks, queues, gen2 blobs, etc?

**Comment by GeeWee** [comment]
Yeah that makes sense - it's definitely a tradeoff, but in a project like this the cost of upgrade isn't that high, so you're right that it might make sense to add in "maybe fixes" as well.

> At what point do you consider the usage high? I'd love to have a target to aim for.
Honestly I wish I had a good answer for you on that one. I think my spider sense is just tingling because it's straigt up using 2-3x as many sockets as anything else.

> For what resources types are you opening tabs? Blobs, disks, queues, gen2 blobs, etc?
I mean, in this case I'm just trying to reproduce the error, but I can give a few examples.
```
Booting up. Sockets used 135
Expanding storage container.  Sockets used 156
(Expanding *multiple* storage containers does not increase socket consumption)
Downloading a file. Sockets 230
(Downloading multiple files does not keep increasing socket consumption)
Expanding a table storage: Sockets 251
Clicking on a table row: Sockets 261
Closing all tabs: Sockets 251
```
Basically it seems like each new functionality loaded takes up a bunch of extra sockets as a one-time cost, but the next time using the functionality, it doesn't keep opening them up.

**Comment by MRayermannMSFT** [comment]
> Honestly I wish I had a good answer for you on that one. I think my spider sense is just tingling because it's straigt up using 2-3x as many sockets as anything else.

Ok, we'll look at some other popular programs to try and get a decent baseline. 

And thanks for the additional numbers! We'll keep looking into this for sure. :)

**Comment by GeeWee** [comment]
Thank you for the quick response!

**Comment by JasonYeMSFT** [comment]
I did a control group testing with our 1.15.1 release version of Storage Explorer and here is my finding:
1. The original socket leakage issue is fixed. Storage Explorer is no longer exhausting the sockets of the system.
2. Our iframe based explorers (e.g. Blob Container, File Share, Table, ADLS Gen1, ADLS Gen2) uses 23 sockets for each type of explorer.
3. Our webview based explorers uses 11 sockets for each explorer.
4. Our settings panel uses 32 sockets.

As long as there isn't too many webview based explorers being opened, Storage Explorer should not run into issue using too many sockets. Given that webview based explorers have their own processes, this kind of limitation is by design.

I would conclude the original issue as fixed. Closing as mitigated.
